### PR TITLE
Implement geometric distribution

### PR DESCRIPTION
**Pull request by xidulu**
https://github.com/pyro-ppl/numpyro/issues/703

As title,

Implementation largely follows torch.distributions.Geometric.

p.s. I know something about torch.distributions but I am relatively new to Jax (still learning the best practice of using Jax.)

### CODE REVIEW

File: numpyro/distributions/util.py [hunk 1]
```diff
@@ -215,6 +215,20 @@ def multinomial(key, p, n, shape=()):
     return _multinomial(key, p, n, n_max, shape)
 
 
+@partial(jit, static_argnums=(2,))
+def _geometric(key, p, shape):
+    if shape is None:
+        shape = jnp.shape(p)
+    dtype = get_dtype(p)
+    tiny = jnp.finfo(dtype).eps
+    u = random.uniform(key, shape, dtype, tiny)
```
Comment (fehiepsi) [inline numpyro/distributions/util.py:224]: nit: because 1 is a valid value for `u`, I guess we can take `u = random.uniform(key, shape, dtype)` and compute `jnp.log1p(-u)` instead of `jnp.log(u)` without having to clamp the value.

File: numpyro/distributions/util.py [hunk 2]
```diff
@@ -215,6 +215,20 @@ def multinomial(key, p, n, shape=()):
...
+    return jnp.floor(jnp.log(u) / jnp.log1p(-p))
```
Comment (fehiepsi) [inline numpyro/distributions/util.py:225]: nit: I think it is better to move this implementation to `sample` method.

File: numpyro/distributions/discrete.py [hunk 3]
```diff
@@ -592,3 +593,76 @@ def mean(self):
...
+    def sample(self, key, sample_shape=()):
+        return geometric(key, self.probs, sample_shape + self.batch_shape)
+
+    @validate_sample
+    def log_prob(self, value):
+        if self._validate_args:
+            self._validate_sample(value)
+        value, probs = promote_shapes(value, self.probs)
+        probs = jnp.where((probs == 1) & (value == 0), 0, probs)
+        return value * jnp.log1p(-probs) + jnp.log(probs)
```
Comment (fehiepsi) [inline numpyro/distributions/discrete.py:651]: You can use [this implementation](https://github.com/pyro-ppl/pyro/blob/dev/pyro/distributions/torch.py#L161) for GeometricLogits. I found it more stable.

File: numpyro/distributions/discrete.py [hunk 4]
```diff
@@ -592,3 +593,76 @@ def mean(self):
...
+    def probs(self):
+        return _to_probs_bernoulli(self.logits)
+
+    def sample(self, key, sample_shape=()):
+        return geometric(key, self.probs, sample_shape + self.batch_shape)
+
+    @validate_sample
+    def log_prob(self, value):
+        if self._validate_args:
+            self._validate_sample(value)
```
Comment (fehiepsi) [inline numpyro/distributions/discrete.py:648]: In NumPyro, you can remove this check (because it is handled in `@validate_sample` decorator).

File: numpyro/distributions/discrete.py [hunk 5]
```diff
@@ -592,3 +593,76 @@ def mean(self):
...
+                                             validate_args=validate_args)
+
+    def sample(self, key, sample_shape=()):
+        return geometric(key, self.probs, sample_shape + self.batch_shape)
+
+    @validate_sample
+    def log_prob(self, value):
+        if self._validate_args:
+            self._validate_sample(value)
+        value, probs = promote_shapes(value, self.probs)
```
Comment (fehiepsi) [inline numpyro/distributions/discrete.py:615]: nit: no need to promote shapes

File: numpyro/distributions/discrete.py [hunk 6]
```diff
@@ -592,3 +593,76 @@ def mean(self):
...
     def variance(self):
         return (1 - self.gate) * self.rate * (1 + self.rate * self.gate)
+
+
+class GeometricProbs(Distribution):
+    arg_constraints = {'probs': constraints.unit_interval}
+    support = constraints.nonnegative_integer
+    is_discrete = True
+
+    def __init__(self, probs=None, validate_args=None):
```
Comment (fehiepsi) [inline numpyro/distributions/discrete.py:603]: nit: `probs=None` -> `probs`

File: numpyro/distributions/discrete.py [hunk 7]
```diff
@@ -592,3 +593,76 @@ def mean(self):
...
+    def variance(self):
+        return (1. / self.probs - 1.) / self.probs
+
+
+class GeometricLogits(Distribution):
+    arg_constraints = {'logits': constraints.real}
+    support = constraints.nonnegative_integer
+    is_discrete = True
+
+    def __init__(self, logits=None, validate_args=None):
```
Comment (fehiepsi) [inline numpyro/distributions/discrete.py:633]: nit: `logits=None` -> `logits`

File: numpyro/distributions/discrete.py (hunk 3 above)
Comment (xidulu) [reply numpyro/distributions/discrete.py:651]: That looks great, didn't find that in the torch repo.

File: numpyro/distributions/util.py (hunk 2 above)
Comment (xidulu) [reply numpyro/distributions/util.py:225]: But does that mean JIT will not be used? I actually follow the practice of other distributions such as Categorical to put the sampling method in util.py.

File: numpyro/distributions/util.py (hunk 1 above)
Comment (xidulu) [reply numpyro/distributions/util.py:224]: I follow the implementation in torch https://github.com/pytorch/pytorch/blob/master/torch/distributions/geometric.py#L78
Also, could `log1p(-u)` be little unstable when `u -> 1`?

Comment (fehiepsi) [reply numpyro/distributions/util.py:224]: In most frameworks, `uniform` samples values in [0, 1), so it should be fine. We can use the trick `1 - u` to get a uniform sample in (0, 1]. If you prefer to sample from (eps, 1), feel free to ignore my comment.

File: numpyro/distributions/util.py (hunk 2 above)
Comment (fehiepsi) [reply numpyro/distributions/util.py:225]: Actually, if you sample from `Geometric`, `random.uniform` will be compiled for you. We often put complicated samplers to `util` to simplify codes in the main implementation. We will move categorical sampler to `CategoricalProbs.sample` when its performance is improved (currently, it is slow when the number of categories are large).

Comment (fehiepsi) [reply numpyro/distributions/util.py:225]: Just a side note: when you move this sampler to GeometricLogits, you can replace `jnp.log1p(-p)` by `-soft_plus(logits)`

File: numpyro/distributions/util.py (hunk 1 above)
Comment (xidulu) [reply numpyro/distributions/util.py:224]: Oh, I got that.
